File: articles/discrete-models-and-recurrence-relations/fortran/recurrence_resource_model.f90
program recurrence_resource_model
  implicit none

  integer :: period, periods
  real(8) :: storage, demand, capacity, inflow, loss_rate, demand_response
  real(8) :: raw_next, shortage, overflow, total_shortage, total_overflow

  storage = 45.0d0
  demand = 10.0d0
  capacity = 80.0d0
  inflow = 4.0d0
  loss_rate = 0.020d0
  demand_response = 0.20d0
  periods = 60
  total_shortage = 0.0d0
  total_overflow = 0.0d0

  do period = 0, periods
    raw_next = storage + inflow - demand - loss_rate * storage
    shortage = max(0.0d0, -raw_next)
    overflow = max(0.0d0, raw_next - capacity)
    total_shortage = total_shortage + shortage
    total_overflow = total_overflow + overflow
    demand = max(0.0d0, demand - demand_response * shortage)
    storage = min(capacity, max(0.0d0, raw_next))
  end do

  print '(A,F12.6,A,F12.6,A,F12.6,A,F12.6)', 'fortran final_storage=', storage, ' final_demand=', demand, ' total_shortage=', total_shortage, ' total_overflow=', total_overflow
end program recurrence_resource_model

File: articles/discrete-models-and-recurrence-relations/cpp/recurrence_resource_model.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

struct Scenario {
    std::string name;
    double initial_storage;
    double initial_demand;
    double capacity;
    double inflow;
    double loss_rate;
    double demand_response;
    int periods;
    bool adaptive_demand;
};

struct Row {
    int period;
    double storage;
    double demand;
    double shortage;
    double overflow;
};

class RecurrenceResourceModel {
public:
    explicit RecurrenceResourceModel(Scenario scenario) : scenario_(std::move(scenario)) {
        if (scenario_.capacity <= 0.0) throw std::invalid_argument("capacity must be positive");
        if (scenario_.initial_storage < 0.0) throw std::invalid_argument("storage must be nonnegative");
        if (scenario_.initial_storage > scenario_.capacity) throw std::invalid_argument("storage exceeds capacity");
    }

    std::vector<Row> simulate() const {
        std::vector<Row> rows;
        double storage = scenario_.initial_storage;
        double demand = scenario_.initial_demand;

        for (int period = 0; period <= scenario_.periods; ++period) {
            const double raw_next = storage + scenario_.inflow - demand - scenario_.loss_rate * storage;
            const double shortage = std::max(0.0, -raw_next);
            const double overflow = std::max(0.0, raw_next - scenario_.capacity);
            const double next_storage = std::min(scenario_.capacity, std::max(0.0, raw_next));

            rows.push_back({period, storage, demand, shortage, overflow});

            if (scenario_.adaptive_demand) {
                demand = std::max(0.0, demand - scenario_.demand_response * shortage);
            }

            storage = next_storage;
        }

        return rows;
    }

private:
    Scenario scenario_;
};

int main() {
    Scenario scenario{"cpp_adaptive", 45.0, 10.0, 80.0, 4.0, 0.020, 0.20, 60, true};
    RecurrenceResourceModel model(scenario);
    const auto rows = model.simulate();

    double total_shortage = 0.0;
    double total_overflow = 0.0;
    for (const auto& row : rows) {
        total_shortage += row.shortage;
        total_overflow += row.overflow;
    }

    std::cout << std::fixed << std::setprecision(6)
              << "cpp final_storage=" << rows.back().storage
              << " final_demand=" << rows.back().demand
              << " total_shortage=" << total_shortage
              << " total_overflow=" << total_overflow
              << "\n";
    return 0;
}
